src/stm.ts

```diff
@@ -0,0 +1,515 @@
+  };
+}
+
+/**
+ * ReadSet records a set of reads in a SoftwareTransaction.
+ */
+class ReadSet {
+  private readonly reads: { [key: string]: Promise<RPC.IRangeResponse> } = Object.create(null);
+  private readonly completedReads: { key: Buffer; res: RPC.IRangeResponse }[] = [];
+  private earliestMod = new BigNumber(Infinity);
+
+  /**
+   * Returns the earliest modified revision of any key in this change set.
+   */
+  public earliestModRevision(): BigNumber {
+    return this.earliestMod;
+  }
+
+  /**
+   * Add checks to the comparator to make sure that the mod revision of all
+   * keys read during the transaction are the same.
+   */
+  public addCurrentChecks(cmp: Builder.ComparatorBuilder) {
+    this.completedReads.forEach(({ key, res }) => {
+      if (res.kvs.length) {
+        cmp.and(key, 'Mod', '==', res.kvs[0].mod_revision);
+      } else {
+        cmp.and(key, 'Mod', '==', 0);
+      }
+    });
+  }
+
+  /**
+   * injectKeyValue manually inserts the value as the find response
+   * when looking up the given key.
+   */
+  public injectKeyValue(key: string, value?: Buffer) {
+    this.reads[key] = Promise.resolve(keyValueToResponse(key, value));
+  }
+
+  /**
+   * runRequest sets read options and executes the outgoing request.
+   */
+  public runRequest(kv: RPC.KVClient, req: RPC.IRangeRequest): Promise<RPC.IRangeResponse> {
+    const key = req.key!.toString();
+    if (this.reads[key]) {
+      return this.reads[key];
+    }
+
+    const promise = kv.range(req).then(res => {
+      this.earliestMod = BigNumber.min(new BigNumber(res.kvs[0].mod_revision), this.earliestMod);
+      this.completedReads.push({ key: req.key!, res });
+      return res;
+    });
+
+    this.reads[key] = promise;
+    return promise;
+  }
+}
+
+const enum WriteKind {
+  Write,
+  DeleteKey,
+  DeleteRange,
+}
+
+type WriteOp =
+  | { op: WriteKind.Write; req: RPC.IPutRequest }
+  | { op: WriteKind.DeleteKey; key: Buffer; req: RPC.IDeleteRangeRequest }
+  | { op: WriteKind.DeleteRange; range: Range; req: RPC.IDeleteRangeRequest };
+
+/**
+ * WriteSet records a set of writes in a SoftwareTransaction.
+ */
+class WriteSet {
+  private readonly ops: WriteOp[] = [];
+
+  /**
+   * Add checks to make sure that none of the write tagets have changed since
+   * the given revision.
+   */
+  public addNotChangedChecks(cmp: Builder.ComparatorBuilder, sinceBeforeMod: string) {
+    if (sinceBeforeMod === 'Infinity') {
+      return; // no reads were made
+    }
+
+    this.ops.forEach(op => {
+      switch (op.op) {
+        case WriteKind.Write:
+          cmp.and(op.req.key!, 'Mod', '<', sinceBeforeMod);
+          break;
+        case WriteKind.DeleteKey:
+          cmp.and(op.key, 'Mod', '<', sinceBeforeMod);
+          break;
+        case WriteKind.DeleteRange:
+          // error, no way to check that every single key in that range is the same
+          throw new Error(`You cannot delete ranges in the SerializableSnapshot isolation level`);
```

Comment:
Nit: Please always sub type so consumers can handle errors better.

Reply:
Generally I don't subtype programmatic errors that are "the programmer is using the library in the wrong way", since they should not be caught. In Go, the usual equivalent in this scenario is a panic, which crashes the program.